contollers/bookings.ts

```diff
@@ -0,0 +1,94 @@
+import { Request, Response } from "express";
+import { bookingType, KeyType } from "../utils/types";
+import emailValidator from "email-validator";
+import moment from "moment";
+
+//Use this array as your (in-memory) data store.
+let bookings: bookingType[] = require("../bookings.json");
+
+const validateBook = (booking: bookingType) => {
+  if (
+    !booking.title ||
+    !booking.firstName ||
+    !booking.surname ||
+    !booking.email ||
+    !emailValidator.validate(booking.email) ||
+    !booking.roomId ||
+    !booking.checkInDate ||
+    !booking.checkOutDate ||
+    !moment(booking.checkOutDate).isAfter(booking.checkInDate)
+  ) {
+    return false;
+  }
+  return true;
+};
+export const getAllBookings = (req: Request, res: Response) => {
+  res.send(bookings);
+};
+
+export const getBooking = (req: Request, res: Response) => {
+  const haveBooking = bookings.find(({ id }) => id === +req.params.bookingsId);
+  if (!haveBooking) {
+    return res.status(404).json(`no booking with id ${req.params.bookingsId}`);
+  }
+  return res.send(haveBooking);
+};
+export const searchBookings = (req: Request, res: Response) => {
+  if (!req.query.term && !req.query.date) return;
+
+  if (req.query.term) {
+    const keys: KeyType[] = ["firstName", "surname", "email"];
+    const filtered = bookings.filter((booking) =>
+      keys.some((key) =>
+        booking[key]
+          .toLowerCase()
+          .includes((<string>req.query.term).toLowerCase())
```

Comment:
@samhoooo Actually, it might be better to change `as string` to `<string>` here?
ps. I didn't expect <string> to be substituted`!` and `.toString()`
Its predecessor was `includes(req.query.term!.toString().toLowerCase())`

Reply:
`as string` is better than `<string>` in here
But if you want more specific typings, you can refer here: https://stackoverflow.com/a/63580201/12219600